src/gan/gen_pyt/asdl/asdl_ast.py

```diff
@@ -0,0 +1,213 @@
+# coding=utf-8
+
+from io import StringIO
+
+from multivac.src.gan.gen_pyt.asdl import *
+
+
+class AbstractSyntaxTree(object):
+    def __init__(self, production, realized_fields=None):
+        self.production = production
+
+        # a child is essentially a *realized_field*
+        self.fields = []
+
+        # record its parent field to which it's attached
+        self.parent_field = None
+
+        # used in decoding, record the time step when this node was created
+        self.created_time = 0
+
+        if realized_fields:
+            assert len(realized_fields) == len(self.production.fields)
+
+            for field in realized_fields:
+                self.add_child(field)
+        else:
+            for field in self.production.fields:
+                self.add_child(RealizedField(field))
+
+    def add_child(self, realized_field):
+        # if isinstance(realized_field.value, AbstractSyntaxTree):
+        #     realized_field.value.parent = self
+        self.fields.append(realized_field)
+        realized_field.parent_node = self
```

Comment:
Not sure I understand what's going on here -- `realized_field.parent_node` is going to inherit the entire class?

Reply:
Each node in the AbstractSyntaxTree is basically defined as a sub-tree, with child nodes as "RealizedFields" - each of which is an AbstractSyntaxTree itself. Each AST remembers its place in the tree with a `parent_node` property that links back to the parent node/AST.